pil-analyzer/src/traits_resolver.rs

```diff
@@ -83,3 +83,32 @@ impl<'a> TraitsResolver<'a> {
         None
     }
 }
+
+pub fn specialize_trait_type<T>(
+    def: &mut Option<FunctionValueDefinition>,
+    trait_type: &Type,
+    named_expr: &mut NamedExpression<Arc<T>>,
+) -> Type {
+    let Some(FunctionValueDefinition::TraitDeclaration(trait_decl)) = def else {
+        panic!("Expected trait declaration");
+    };
+
+    let Type::Tuple(TupleType { items }) = trait_type else {
+        panic!("Expected tuple type for trait implementation");
+    };
+
+    let type_var_mapping: HashMap<String, Type> = trait_decl
+        .type_vars
+        .iter()
+        .cloned()
+        .zip(items.iter().cloned())
+        .collect();
+
+    let trait_fn: &mut TraitFunction = trait_decl
+        .function_by_name_mut(&named_expr.name)
+        .expect("Function not found in trait declaration");
+
+    trait_fn.ty.substitute_type_vars(&type_var_mapping);
```

Comment:
Ah, now I see! Does it even have a type? Shouldn't we just take the type from the trait and then let the type inferenec algorithm do its job?

Reply:
I.e. just return `trait_type.clone().substitute_type_vars(&type_var_mapping)`?

Reply:
And then I think the type inference algorithm will store the type anyway at the end.